crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Simple upgrade policy that requires a `k` out of `n` quorum in order to perform
+/// a proposed upgrade.
+/// 
+/// This policy is initiated with a call to `k_of_n_upgrade_policy::new` providing 
+/// the `UpgradeCap` of the package to be controlled by the policy, the `k` value 
+/// (number of votes to be received for the upgrade to be allowed) and the list of
+/// `address`es allowed to vote. The `address`es provided will receive
+/// a `Ballot` for any given proposal. The sender of the transaction that creates
+/// the `KofNUpgradeCap` is added to the list of voters as well.
+/// The set of possible voters is provided at creation time and it is immutable for
+/// the lifetime of the policy.
+/// This policy allows for a number of voters between 2 and 100.
+/// 
+/// An upgrade is proposed via `k_of_n_upgrade_policy::propose_upgrade` and the sender
+/// of the upgrade not only needs to have a reference to the `KofNUpgradeCap` but they 
+/// must also be in the list of possible voters.
+/// A successful call for a valid proposal will result in all `address`es
+/// registered to receive a `Ballot` which would allow them to vote for the proposal.
+/// The proposer will not receive a `Ballot`.
+/// Receiver of the `Ballot` can then vote for the upgrade. The `Ballot` can also
+/// be sent to other entities which can then vote for the upgrade on behalf of
+/// the original owner.
+/// 
+/// Once the number of votes is reached the proposer of the upgrade can perform
+/// the upgrade.
+/// 
+/// Events are emitted to track the main operations on the proposal.
+/// A proposed upgrade lifetime is tracked via the 4 events:
+/// `UpgradeProposed`, `UpgradeVoted` and `UpgradePerformed` or `UpgradeDiscarded`.
+/// Also `Ballot`s are tracked via `BallotTransfered` and `BallotDeleted` in order to
+/// audit the lifetime of a vote.
+/// 
+/// Multiple upgrades can be live at the same time. That is not the expected behavior
+/// but there are no restriction to the number of upgrades open at any point in time.
+/// When that happens the first upgrade executed "wins" and subsequent attempt to
+/// authorize an upgrade will fail as the version will not match any longer.
+module sui::k_of_n_upgrade_policy {
+    use std::vector;
+    use sui::event;
+    use sui::object::{Self, ID, UID};
+    use sui::package::{Self, UpgradeCap, UpgradeTicket, UpgradeReceipt};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+    use sui::vec_set::{Self, VecSet};
+
+    /// The capability controlling the upgrade. 
+    /// Initialized with `new` is returned to the caller to be stored as desired.
+    struct KofNUpgradeCap has key, store {
```

Comment:
nit, should this be `KOfNUpgradeCap`.. oh god thats kind of bad too. 

Is there a canonical short name for "k of n"?

Reply:
if you are questioning the name please help out because yes I agree... on whatever that was....

Reply:
I think Todd doesn't like the capitalisation mainly 😅

Reply:
got it :)

Reply:
`QuorumUpgradeCap`, maybe (shrinks from 3 syllables down to 2...) 😆

Reply:
we are definitely going to change the name and so far I am liking `QuorumUpgradeCap` though other names welcome.
I will change the code once we decide how to ship this (framework or not)

Reply:
`QuorumUpgradeCap` sounds good to me too!